Preview -  Flink 5734 implement code generation for normalizedkey sorter

p16i opened the pull request.

greghogan committed:
[FLINK-3722] [runtime] Don't / and % when sorting

Replace division and modulus with addition and subtraction.

greghogan committed:
Additional comments

p16i committed:
[FLINK-5734] prepare project structure

p16i committed:
[FLINK-5734] add sorter template

p16i committed:
[FLINK-5734] generate swap procedures

p16i committed:
[FLINK-5734] add getByte/putByte

p16i committed:
[FLINK-5734] add test cases.

p16i committed:
[FLINK-5734] use singleton pattern

p16i committed:
[FLINK-5734] rearrange files

p16i committed:
[FLINK-5734] pass tests for compare optimization

p16i committed:
[FLINK-5734] remove tap from generated code for readability

p16i committed:
[FLINK-5734] use NormaliKeysorter's appoach to compute numKeyByte

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterFactory.java:
```diff
@@ -0,0 +1,73 @@
+		);
+	}
+
+	public static SorterFactory getInstance() throws IOException {
```
I think we have a thread-safety issue here. Maybe make this `synchronized`? (And the same in `TemplateManager`)

p16i replied:
You're right. I forgot to consider that. I will go with `double-checked locking` approach.
https://en.wikipedia.org/wiki/Double-checked_locking

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,276 @@
+	private final ArrayList<Integer> byteOperators;
+	private final String sorterName;
+	private final int numBytes;
+	private final boolean isKeyFullyDetermined;
```
comment please

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,276 @@
+	}
+
+	public boolean isSortingKeyFixedSize(){
+		System.out.println(typeComparator.getNormalizeKeyLen());
```
Use logging instead.

p16i replied:
actually, i just realized that we don't use this method. I will remove it.

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,276 @@
+public class SorterTemplateModel {
+	public final static String TEMPLATE_NAME = "sorter.ftlh";
+
+	private final static Integer[] POSSIBLE_FIXEDBYTE_OPERRATORS = {8,4,2,1};
```
Typo: opeRRators

p16i replied:
nice spot!

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,276 @@
+					offsetString = "+" + accOffset;
+				}
+
+				temporaryString     += String.format("%s temp%d = segI.get%s(iBufferOffset%s);\n",primitiveType, varIndex, primitiveClass, offsetString );
```
`+=` is slow on String because it is immutable, so it copies the whole thing. Please use `StringBuilder` instead.

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,276 @@
+			for( int i = 1; i < byteOperators.size(); i++ ){
+				int noBytes = byteOperators.get(i);
+				if( noBytes == 1 ){
+					break;
```
This depends on that the 1-byte chunk can only be the last, right? Could you change this into `continue`? It might happen that someone sometime changes the decomposition and then it's a bug.

Or, alternatively, you could write a comment to the decomposition, saying that it is important that the 1-byte chunk comes last.

p16i replied:
actually, `1-byte` always comes last because `byteOperators` is computed using greedy approach from `POSSIBLE_FIXEDBYTE_OPERATORS`.

ggevay replied:
OK, if you don't change it to `continue`, then please write a comment here saying that `break` is OK here because ... (so that people reading the code don't have to spend time thinking about this), and also a comment to the computation of `byteOperators` saying that it is important that the 1-byte chunk comes last (so that someone changing the logic there won't accidentally break this).

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/TemplateManager.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.Map;
+
+public class TemplateManager {
+	public static String RESOURCE_PATH  = "/Users/heytitle/projects/apache-flink/flink-runtime/resources";
```
change it to a relative path

p16i replied:
@SerkanAli could you please change this to use Flink's temporary directory instead?

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/TemplateManager.java:
```diff
@@ -0,0 +1,88 @@
+		String generatedFilename = model.getSorterName();
+
+		if( generatedSorter.getOrDefault(generatedFilename, false) ){
+			System.out.println("Serve from cache!");
```
log instead

ggevay commented on flink-runtime/src/test/java/org/apache/flink/runtime/codegeneration/SortingTest.java:
```diff
@@ -0,0 +1,617 @@
+
+			final int cmp = last.compareTo(current);
+			if (cmp > 0) {
+				Assert.fail("Next value is not larger or equal to previous value.");
```
Please also test that the same numbers appear as those that were written, and that all of them appears.

p16i replied:
I have a question about this. I'm thinking to have one array containing expected order sorted by Java sort function and the other one from generated sorter.  I plan to use `Assert.assertArrayEqual` to verify the result.

However,  it might have a situation that 2 records have the same key and the orders are slight different due to the fact that Quicksort is not a stable sort algorithm, hence the assertion returns `false`.

Do you have any better approach how to compare these 2 arrays?

ggevay replied:
I can't think of any simple way to do this. `Assert.assertArrayEqual` doesn't work because of the reason that you mentioned.

I think you should just check that the final order is sorted, and that the same elements appear the same number of times.

ggevay commented on flink-runtime/src/test/java/org/apache/flink/runtime/codegeneration/SortingTest.java:
```diff
@@ -0,0 +1,617 @@
+		InMemorySorter<Tuple2<Integer, String>> sorter = createSorter(TestData.getIntStringTupleSerializer(), accessors, memory);
+
+		TestData.TupleGenerator generator = new TestData.TupleGenerator(SEED, KEY_MAX, 5, KeyMode.RANDOM,
+			ValueMode.FIX_LENGTH);
```
Is this testing the situation when the string is longer that the normalized key? (i.e., when the `else` branches are hit in the not fully determined case)

p16i replied:
nope. I will add another test case for that.

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,276 @@
+
+			}
+		} else {
+			procedures += "";
```
Is this case still a TODO, or am I missing something?

p16i replied:
I think it is unnecessary. I'll remove it.

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,276 @@
+	private String generateSwapProcedures(){
+		String procedures = "";
+
+		if( this.byteOperators.size() > 0 ) {
```
When does it happen that this is false?

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,276 @@
+
+	private final HashMap<Integer,String> byteOperatorMapping;
+	private final TypeComparator typeComparator;
+	private final ArrayList<Integer> byteOperators;
```
Maybe you could give a better name here. For example, `chunkSizes`?

p16i replied:
ok, that sounds better.

ggevay reviewed (changes requested):
Generally, it looks good! I wrote a few minor comments inline.

Please write more comments in the code. (It's mostly OK for me, because we had many discussions about the topic, but someone who doesn't know the topic will have a hard time understanding the code without comments.)

ggevay commented:
Btw. you could add a manual performance test for sorting, like `flink/flink-tests/src/test/java/org/apache/flink/test/manual/ReducePerformance.java`. This would make later experimenting with further performance improvements easier.

p16i committed:
[FLINK-5734] rename variable names

p16i committed:
[FLINK-5734] use synchronized block for SorterFactory and TemplateModel

p16i committed:
[FLINK-5734] remove unused code

p16i committed:
[FLINK-5734] replace string concat with stringbuilder

p16i committed:
[FLINK-5734] add testcase for variable-length string

p16i committed:
[FLINK-5734] user proper logger

p16i committed:
[FLINK-5734] add more comments

p16i committed:
[FLINK-5734] checking endianness should not be involved in generating procedures for compare function

p16i committed:
[FLINK-5734] move byte-operator-mapping to be a constant of class scope

p16i committed:
[FLINK-5734] add enable/disable flag in ExecutionConfig

p16i committed:
[FLINK-5734] fix tests of SorterTemplateModel due to change in computing numKeyBytes

p16i committed:
[FLINK-5734] rename variable names in old compare/swap functions to match with the new ones

p16i committed:
[FLINK-5734] refactor tests

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/Somethig.java:
```diff
@@ -0,0 +1,12 @@
+//	public String toString() {
+//		return super.toString();
+//	}
+//}
```
delete

ggevay commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterTemplateModel.java:
```diff
@@ -0,0 +1,342 @@
+	 * Getter for sorterName which instantiated from constructor
+	 * @return name of the sorter
+	 */
+	public String getSorterName (){
```
inconsistent whitespace

p16i replied:
sry, I don't see the issue you are mentioning.

ggevay replied:
It's just that there is a space before `()` but not after, and in other methods the space is after `()`.

p16i committed:
[FLINK5743] remove unused file

p16i committed:
[FLINK-5734] update comment

p16i committed:
[FLINK-5734] also tests that no record is discarded after sorting

p16i committed:
[FLINK-5734] add timestamp to generated sorter

p16i committed:
[FLINK-5734] test TemplateManager caching mechanism

p16i commented on flink-runtime/resources/templates/sorter.ftlh:
```diff
@@ -35,6 +35,9 @@ import java.util.List;
 import org.apache.flink.runtime.operators.sort.LargeRecordHandler;
 
+/**
+ * This code is generated on ${.now}
```
@ggevay do you where can I get Flink's `version? I think it will be useful for debugging if the generated code also contains the version.

ggevay replied:
I don't know unfortunately.

p16i committed:
[FLINK-5734] fix coding style

SerkanAli committed:
[FLINK-5734] RESOURCE_PATH exhibits to a Temporary directory

p16i commented on flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/TemplateManager.java:
```diff
@@ -41,10 +41,10 @@
-	public static final String RESOURCE_PATH  = "/Users/heytitle/projects/apache-flink/flink-runtime/resources";
-	public static final String TEMPLATE_PATH  = RESOURCE_PATH + "/templates";
+	public static final String RESOURCE_PATH  = createTemporaryDirectory();
+	public static final String TEMPLATE_PATH  = createSubFolder(RESOURCE_PATH , "templates");
```
`TEMPLATE_PATH` should link to `resources` directory where `sorter.fthl` locates, not the temporary directory.  On the other hand, `GENERATING_PATH` should link to temporary directory and I think we should try to use the one that Flink use.

p16i committed:
[FLINK-5734] rename variable names

ggevay commented:
What is the relation between the new tests that you wrote, and any old tests that are testing the old sorting code (e.g. `SortPartitionITCase`)? Are the new tests completely separate? Maybe it would it make sense to  also modify some of the old tests to run them with twice, once without code generation and once with code generation.

heytitle commented:
@ggevay I will look into that test.

Also, I have a question regarding integrating this code to Flink. One of the classes that uses `NormalizedKeySorter` is `UnilateralSortMerger`. This class doesn't know execution config, so that means we have to pass the config when instantiating an object of `UnilateralSortMerger` and it seems to be instantiated many places.

Do you know any simple approach to make this work, otherwise I have to make changes to those places?

ggevay commented:
> `UnilateralSortMerger`. This class doesn't know execution config

You can actually access the `ExecutionConfig` in `UnilateralSortMerger`: The constructor gets the `parentTask` as an `AbstractInvokable`, from which you can ask the `Environment`, from which you can ask the `ExecutionConfig`.

SerkanAli committed:
[FLINK-5734] Change RESOURCE_PATH to the Working place

SerkanAli committed:
Merge branch 'preview-FLINK-5734-implement-code-generation-for-normalizedkey-sorter' of https://github.com/heytitle/flink into preview-FLINK-5734-implement-code-generation-for-normalizedkey-sorter

p16i committed:
[FLINK-5734] create directory for generated if it doesn't exist.

Also, change the way generated code loaded because the old approach
throws ClassNotFound exception when multi file are created at once.

heytitle commented:
Thanks for pointing out. I literally didn't see that.

p16i committed:
[FLINK-5734] integrate code generation to flink code

p16i committed:
[FLINK-5734] move templates to proper location

p16i committed:
[FLINK-5734] prevent generating same sorter at the same time

p16i committed:
[FLINK-5734] be able to run SortPartitionITCase with and without code generation

p16i committed:
[FLINK-5734] remove debug line

p16i committed:
[FLINK-5734] fix sorting in desc order failed

p16i committed:
[FLINK-5734] remove unused imports

p16i committed:
[FLINK-5734] update toString of custom sorter

p16i committed:
[FLINK-5734] also cache contructor to save cooking time

p16i committed:
[FLINK-5734] refactor integration tests

p16i committed:
[FLINK-5734] remove unused imports

p16i committed:
[FLINK-5734] add benchmarking code

p16i committed:
[FLINK-5734] fix test due to change in naming

SerkanAli committed:
[FLINK-5734] change the Temporary directory defined by the ConfigConstant

p16i committed:
[FLINK-5734] remove unused import

p16i committed:
Merge branch 'preview-FLINK-5734-implement-code-generation-for-normalizedkey-sorter' of github.com:heytitle/flink into preview-FLINK-5734-implement-code-generation-for-normalizedkey-sorter

p16i committed:
[FLINK-5734] correct resource path

ggevay commented:
You should squash together most of your commits before opening the pull request to the Flink repo. However, if you can group your changes into 2-4 logical units, then having separate commits for these logical units improves the readabilty of the pull request a lot, and increases the chances that someone will review it soon.

ggevay commented:
> This is always return the default value because you retrieve the value from new configuration object which is not the one from Flink job.

Hm, you mean this for the config that you access from `UnilateralSortMerger`? This would be strange.

heytitle commented:
@ggevay if you look at the code, the configuration object is newly instantiated. It isn't the once from `UnilateralSortMerger`. 

REF: https://github.com/heytitle/flink/commit/651ca01ac62dbc85daa1e6c4d8ebcbfa26f20d9f#diff-2a2c65c22d4776de043e0198530f3ab3R138

SerkanAli committed:
[FLINK-5734] Change empty config to the GlobalConfiguration
